src/arrays.jl

```diff
@@ -4,23 +4,49 @@ const UniFinArr = UnivariateFiniteArray
 
 Base.size(u::UniFinArr, args...) =
     size(first(values(u.prob_given_ref)), args...)
-
-function Base.getindex(u::UniFinArr{<:Any,<:Any,R,P,N},
-                       i::Integer...) where {R,P,N}
-    prob_given_ref = LittleDict{R,P}()
-    for ref in keys(u.prob_given_ref)
-        prob_given_ref[ref] = getindex(u.prob_given_ref[ref], i...)
+    
+function Base.getindex(u::UniFinArr{<:Any,<:Any,R, P}, i...) where {R, P}
+    # It's faster to generate `Array`s of `refs` and indexed `ref_probs`
+    # and pass them to the `LittleDict` constructor.
+    # The first element of `u.prob_given_ref` is used to get the dimensions
+    # for allocating these arrays.
+    u_dict = u.prob_given_ref
+    a, rest = Iterators.peel(u_dict)
+    # `a` is of the form `key => value`. 
+    a_ref, a_prob = first(a), getindex(last(a), i...)
+    
+    # Preallocate Arrays using the key and value of the first 
+    # element (i.e `a`) of `u_dict`. 
+    n_refs = length(u_dict)
+    refs = Vector{R}(undef, n_refs)
```

Comment:
Your approach allocates `refs` and `ref_probs` arrays. The design where `i::Integer...` and `general `I...` (and also what I proposed in #43) were separated avoided such allocation. Have you benchmarked that what you propose does not lead to significant regression (in time and in number of allocations). This is especially relevant for `i::Integer...` case.

Reply:
```julia
julia> using CategoricalDistributions, BenchmarkTools

julia> function getindex2(u::UnivariateFiniteArray{<:Any,<:Any,R,P,N},
                              I...) where {R,P,N}
           prob_given_ref = nothing
           for ref in keys(u.prob_given_ref)
                   idx_res = getindex(u.prob_given_ref[ref], I...)
               if isnothing(prob_given_ref)
                    # make this check just to be sure that nothing unexpected happened
                   @assert idx_res isa Array{P}
                   prob_given_ref = CategoricalDistributions.LittleDict{R,Array{P,ndims(idx_res)}}()
               end
               prob_given_ref[ref] = idx_res
           end
           return UnivariateFiniteArray(u.scitype, u.decoder, prob_given_ref)
       end
getindex2 (generic function with 2 methods)

julia> function getindex2(u::UnivariateFiniteArray{<:Any,<:Any,R,P,N},
                              i::Integer...) where {R,P,N}
           prob_given_ref = CategoricalDistributions.LittleDict{R,P}()
           for ref in keys(u.prob_given_ref)
               prob_given_ref[ref] = getindex(u.prob_given_ref[ref], i...)
           end
           return UnivariateFinite(u.scitype, u.decoder, prob_given_ref)
       end
getindex2 (generic function with 2 methods)

julia> u4 = UnivariateFinite(['x', 'y', 'z'], rand(10000, 4, 3), pool=missing, ordered=true);

julia> u40 = UnivariateFinite(['x', 'y', 'z'], rand(10000, 40, 3), pool=missing, ordered=true);

julia> u400 = UnivariateFinite(['x', 'y', 'z'], rand(10000, 400, 3), pool=missing, ordered=true);
julia> @btime getindex($u4, 238, 3)
  54.268 ns (2 allocations: 144 bytes)
         UnivariateFinite{OrderedFactor{3}}
     ┌                                        ┐
   x ┤■■ 0.039016
   y ┤■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 0.612437
   z ┤■■■■■■■■ 0.167377
     └                                        ┘

julia> @btime getindex2($u4, 238, 3)
  389.109 ns (13 allocations: 464 bytes)
         UnivariateFinite{OrderedFactor{3}}
     ┌                                        ┐
   x ┤■■ 0.039016
   y ┤■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 0.612437
   z ┤■■■■■■■■ 0.167377
     └                                        ┘

julia> @btime getindex($u40, 238, 3)
  58.189 ns (2 allocations: 144 bytes)
         UnivariateFinite{OrderedFactor{3}}
     ┌                                        ┐
   x ┤■■ 0.0558472
   y ┤■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 0.770212
   z ┤■■■■■■■■■■■■■■■■■■■■■■■■ 0.60511
     └                                        ┘

julia> @btime getindex2($u40, 238, 3)
  390.594 ns (13 allocations: 464 bytes)
         UnivariateFinite{OrderedFactor{3}}
     ┌                                        ┐
   x ┤■■ 0.0558472
   y ┤■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 0.770212
   z ┤■■■■■■■■■■■■■■■■■■■■■■■■ 0.60511
     └                                        ┘
julia> @btime getindex($u400, 238, 3)
  56.358 ns (2 allocations: 144 bytes)
         UnivariateFinite{OrderedFactor{3}}
     ┌                                        ┐
   x ┤■■■■■■■■ 0.221749
   y ┤■■■■■■■■■■■■■■■■■■■■■■ 0.590628
   z ┤■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 0.807161
     └                                        ┘

julia> @btime getindex2($u400, 238, 3)
  392.079 ns (13 allocations: 464 bytes)
         UnivariateFinite{OrderedFactor{3}}
     ┌                                        ┐
   x ┤■■■■■■■■ 0.221749
   y ┤■■■■■■■■■■■■■■■■■■■■■■ 0.590628
   z ┤■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 0.807161
     └                                        ┘
julia> @btime getindex($u4, 78:9000, 1:3);
  49.300 μs (8 allocations: 627.84 KiB)

julia> @btime getindex2($u4, 78:9000, 1:3);
  55.300 μs (36 allocations: 628.91 KiB)

julia> @btime getindex($u40, 78:9000, 1:15);
  537.300 μs (8 allocations: 3.06 MiB)

julia> @btime getindex2($u40, 78:9000, 1:15);
  549.200 μs (36 allocations: 3.06 MiB)

julia> @btime getindex($u400, 78:9000, 1:150);
  5.911 ms (8 allocations: 30.64 MiB)

julia> @btime getindex2($u400, 78:9000, 1:150);
  5.935 ms (36 allocations: 30.64 MiB)

julia> h = UnivariateFinite(['x', 'y', 'z', 'a', 'b', 'c', 'd', 'e', 'f', 'g'], rand(10000, 4, 10), pool=missing, ordered=true);

julia> @btime getindex($h, 350, 2);
  65.577 ns (2 allocations: 208 bytes)

julia> @btime getindex2($h, 350, 2);
  990.000 ns (29 allocations: 1.06 KiB)

julia> @btime getindex($h, 78:9000, 1:3);
  187.400 μs (22 allocations: 2.04 MiB)

julia> @btime getindex2($h, 78:9000, 1:3);
  197.600 μs (73 allocations: 2.05 MiB)

julia> g = UnivariateFinite(["x$i" for i in 1:20], rand(10000, 4, 20), pool=missing, ordered=true);


```
@bkamins the new implementation seems slightly faster than the existing one. The thing is that creating an empty `LittleDict` allocates two new `Array`s (`keys` and `values`, both initially empty). Then calling something like `prob_given_ref[ref] = probs` would push to the `keys` and `values`. As the number of classes increases, pushing to these arrays would require more allocations for obvious reasons. Julia's seems very good at pre-allocating.

Reply:
So `LittleDict{R, eltype(ref_probs)}(refs, ref_probs)` does reuse `refs` and `ref_probs` without copying them? Nice. Then your approach is indeed much better.